_episodes/01-numpy.md

```diff
@@ -57,7 +57,7 @@ weight_kg = 60
 {: .language-python}
 
 From now on, whenever we use `weight_kg`, Python will substitute the value we assigned to
-it. In essence, **a variable is just a name for a value**.
+it. In essence, **a variable is a name for a value**.
```

Comment:
My suggestions are:

> Basically, **a variable is a name for some value**.

> Basically, **a variable (`weight_kg`) is a name for a value (`60`)**.

> In layman's terms, **a variable is a name for a value**.

What do you think, @n2ygk, @brownsarahm, @ldko?

Reply:
Since we are looking at this closely, I think any of these could be considered misleading because the variable is not just the name, but the stored value with a name. This difference is highlighted with the next line `In Python, variable names` where "variable names" is used to indicate the name of a variable over just "variable". Because of that, my suggestion is to say something like:
```
From now on, wherever we use `weight_kg`, Python will substitute the value we assigned to it. Basically, a variable is a value that has a name.
```

Reply:
This is a tricky one, because what it's doing is explicitly not addressing what actually happens and tries to suggest a heuristic as a mental model. I personally think that a name for a value is a bad mental model for a variable, because it suggests permanence. However, I agree that we want to introduce some heuristic and not get into actually how memory works. Maybe an appropriately simplified version is that a variable is a named container that we can put values in and change. 

```
One way to think of a variable is that it is a name for a container and when we refer to it, it's contents, a value is filled in by python. 
```

I think for the carpentries' target audience of researchers, the concept of a variable in algebra is reliable prior knowledge. So we could instead build from there. I think this approach is better, it avoids demotivating language and tries to use something learners are more familiar with to teach the new concept.  

```
Like a variable in a mathematical expression, a variable stands in for a value.  What's different is that variables in python can have a wide variety of values and the value of a variable can change. 
```

Reply:
I prefer the explanation of a variable of a container. This extends well to further explaining that the value is the content of the container, but code can be developed without having that content, values are only required at run time.

I'm less sure about the mathermatical reference. I don't understand how variables taking a wide variety of values is "different" from a variable in a mathematical expression, as a mathematical variable can take any value in e.g. a function's domain. And more generally, it's the nature of a variable that the values it holds can vary.

Reply:
I like the track of a variable being described as a named container. I also would prefer to avoid bringing in the mathematical expression comparison.

Reply:
Thank you all, for your comments! This is indeed a tricky question/concept to explain. Here are my thoughts...

TL;DR

Let's discuss this sentence/paragraph/section in a separate PR/issue.

<hr>

> Basically, a variable is a value that has a name.

This definition emphasizes value over name and, therefore, is even more vulnerable to criticism of implying a static nature of value that @brownsarahm mentions as a critique of the current definition. I also think this definition changes the meaning of the sentence and of what a variable is. Here are some analogies:

<br>

(Current version): "In essence, Buddy (variable) is a nickname for a dog (value)."
(This suggestion): "Basically, Buddy (variable) is a dog (value) that has a nickname."

(Current version): "In essence, a nickname (variable) is a [short] name for a person (value)."
(This suggestion): "Basically, a nickname (variable) is a person (value) with a [short] name."

(Current version): "In essence, a laptop (variable) is a [short] name for a small computer (value)."
(This suggestion): "Basically, a laptop (variable) is a small computer (value) that has a [short] name."

<br>

<hr>

> the variable is not just the name, but the stored value with a name.

This is where it becomes tricky. Usually (and by saying "usually" I mean languages like C, Fortran), a variable is a human-readable name for an address in (and the required size of) computer memory. This memory address always holds some value:

> Human --> variable --> memory address (+ size) === value

`-->` means abstractions
`===` means equivalence.

So, usually, the connection between a variable and a value is through a memory address that does not change once a variable is created. For example, one can create an integer `int a = 0;` in C, check its address in memory, increment value by 1 with `a=a+1;`, and address of `a` in memory will be the same. In Python, even simple operations like `a=a+1` change memory address of a variable. Given that both value and memory address can change, variable's name is, in fact, the only thing that is permanent. That why making a mental connection between variable and its name is probably the right thing to do in Python.

What I don't "like" about this definition is the word "value": it is too abstract and it indeed might look as if we're giving a name to a specific value. Perhaps, "observable quantity" or "data" are better alternatives to "value".

<hr>

### Side notes

The most important thing about variables is that they are human-readable names and not something like `0x7ffee16c26fc`. So, while variables are certainly connected to values, they are not values themselves. 

In Python, variables cannot exist without values (`None` is also a value). In other languages they can: `int a;` in C declares `a` as a name for a variable of type integer. However, before setting its value with something like `a = 10;` the value of `a` is not defined -- it (the value) depends on what is located at the memory address that operating system gives to the program at the time someone asks about the value of `a`.

<hr>

> This is a tricky one, because what it's doing is explicitly not addressing what actually happens and tries to suggest a heuristic as a mental model. I personally think that a name for a value is a bad mental model for a variable, because it suggests permanence.

I agree that "a name for a value" is not the best definition but it's the word "value" that is not right here. See my notes above. How about, "observable quantity" or "data"?

<hr>

> Maybe an appropriately simplified version is that a variable is a named container that we can put values in and change.

> I prefer the explanation of a variable of a container. This extends well to further explaining that the value is the content of the container, but code can be developed without having that content, values are only required at run time.

> I like the track of a variable being described as a named container.

I don't mind introducing this mental model and, perhaps, we should. Note, however, that we already have one mental model -- variables as sticky notes -- which is not a bad one. Perhaps, we could extend / amend / clarify it. I like this definition of variables: https://launchschool.com/books/ruby/read/variables#whatisavariable

Reply:
Maybe I'll just borrow this definition: http://swcarpentry.github.io/python-novice-gapminder/02-variables/index.html. I'm not going anywhere near launchschool.com's definition as it is not CC licensed.

Reply:
... which means leaving it as is.